### ISSUE TITLE
[BlenderBIM] Interface strings should be i18n

### ISSUE THREAD

**Issue by globalcitizen** [IfcOpenShell/IfcOpenShell#889]
Strings should be internationalized. 

[Blender's notes on internationalization](https://wiki.blender.org/wiki/Source/Interface/Internationalization) has specific notes on [translating unofficial addons](https://wiki.blender.org/wiki/Process/Translate_Blender#Translating_non-official_addons).

**Comment by Moult** [comment]
This is a valid bug, but a very low priority, considering that UI changes are very frequent right now.

If you'd like to translate a few of the more stable parts of the UI, go for it. Once you've done a small example, this can be labeled a good first issue for future contributors.

**Comment by Moult** [comment]
A year later, and the UI is still changing too frequently for this bug to be considered a high priority.

**Comment by Andrej730** [comment]
Got some progress to share on this. I've investigated the blender way for addons localization. The main way is to setup translation dictionary (similar to the one below) and register it with `bpy.app.translations.register(__name__, translations_dict)`. Then Blender automatically replaces UI strings bases on current locale.

```python
    (("*", "Decorations Colour"),
     (("bpy.types.BIM_ADDON_preferences.decorations_colour",),
      ()),
     ("de_DE", "",
               (False, ())),
     ("ru_RU", "",
               (False, ())),
    ),
```

Creating these dictionaries is automated using official Blender addon "Manage UI Translations" - it automatically parses strings from addon's Blender structures and creates `translation_dict` saved to module `translations.py`. There are some caveats I'm trying to workaround right now ([found some bug in Blender](https://projects.blender.org/blender/blender/issues/116579)) but ideally it will get all those strings to translate from the code automatically.

Then same addon has option to save translation to `.po` files (1 .po file per language) which are much more readable than `translations.py` and anyone would be able to edit them and add translated versions of strings. Then this addon has option to import .po files back to `translation.py`. Basically `translation.py` is a container Blender uses to register the translation strings and it's maintained automatically.

Example of .po file:
```po
#. :src: bpy.types.LOCALIZATION_TEST_OT_test
msgctxt "Operator"
msgid "Operator name"
msgstr ""

#. :src: bpy.types.LOCALIZATION_TEST_OT_test
msgid "Operator description"
msgstr ""

#. :src: bpy.types.LOCALIZATION_TEST_PT_panel
msgid "Localization test"
msgstr ""

#. :src: bpy.types.LOCALIZATION_TEST_PT_panel
#, fuzzy
msgid "Panel Header Updated"
msgstr "Test panel header RU"
```

I've created some test addon to show an example - https://github.com/Andrej730/localization_test/commits/master/

Ping @Gorgious56 as you might be interested in funky blender stuff 😁

**Comment by aothms** [comment]
This surely is next-level ground breaking stuff in making the UI even more accessible. Great work.

I just wanted to let you know we also have translation for IFC in PO files. https://github.com/buildingSMART/IFC4.3.x-output/tree/translations/translations The location (git repo/branch) of these might change, but it might be an easy way to change the IFC'ish term IfcWall to e.g  دیوار pared  mur  Wand  muro  parede  muur  стена  壁  墙  جدار  दीवार  벽  τοίχος  duvar  vägg

**Comment by Andrej730** [comment]
I keep investigating, I'll put notes I've got so far on how it would work.

There is a need to have `translations.py` right besides BBIM `__init__.py` because Blender translation addons expects it to be there to convert it to .po
File example - https://github.com/Andrej730/localization_test/blob/cc9566fa6b8299812222a640a1e035a2926245af/translations.py

There also will be bunch of `.po` files - they have simpler format, so they will be actually used to translate strings.
File example - https://github.com/Andrej730/localization_test/blob/cc9566fa6b8299812222a640a1e035a2926245af/de_DE.po

There is no need to store `.po` files in BlenderBIM folder, they will be just used to generate `translations.py` by Blender translation addon.
Noticed that Blender itself had a similar issue, so they just store them in a separate repository to keep the commit history clean.

So, my current idea is that there will be a second repository like `blenderbim-ui-translations` that will store `translations.py` and will have `locale` folder with `.po` files.
As `translations.py` will be used in BlenderBIM but not included in IfcOpenShell, for users it will be included to builds from `blenderbim-ui-translations`.
As for developers, they can either manually copy `translations.py` if they want to update it (like we do with `ifcopenshell_wrapper.pyd`) or to receieve most recent translations they'll need to clone `blenderbim-ui-translations` and symlink `translations.py` to their BlenderBIM addon folder.

Workflow for users who would want to commit the translated strings:
`-- steps that need to be done only once --`
1) clone `blenderbim-ui-translations`
2) symlink `translations.py` and `locale` folder to `addons/blenderbim`
3) enable "Manage UI Translation" addon
4) setup necessary paths for that addon (I18n branches, source root, translation root) - all can be just empty directories, as we just need them to get addon stop complaining

`------------------------------------------`

5) setting up i18n branches has extra step of calling `setup_translations.setup_branches()` to rearrange .po ready to translation
6) translate strings - either by manually editing .po files or right from UI - https://i.imgur.com/sJj5qsq.png
7) run `setup_translations.update_translations_from_po()` to update `translations.py`
8) commit
Ideally, all those steps will be automated and those function calls can be exposed to UI, so user will just use two buttons like "Setup translation environment" and "Update translation data" and then will commit them to git.

There also will be a workflow to update `translations.py` and `.po` files in case BlenderBIM got new strings:
1) disable BlenderBIM addon and restart Blender
(currently there is an issue with Blender where it fails to parse some string if parsed addon is enabled at beginning of the parsing)
2) `setup_translations.reload_translations()` - will create updated `translations.py`
3) `setup_translations.convert_translations_to_po()` - will update .po files

**Comment by Moult** [comment]
BTW - https://github.com/buildingSMART/IFC4.3.x-development/issues/763

**Comment by Andrej730** [comment]
I've set up translations in current BlenderBIM builds generated daily in https://github.com/IfcOpenShell/IfcOpenShell/releases.

Explanation how it works for everyone involved in translations.

### Basic setup
1. Clone translations repo https://github.com/IfcOpenShell/blenderbim-translations and install BBIM Translations addon from https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/scripts/bbim_translations.py.
2. Start Blender
3. In Blender properties choose the `blenderbim-translations` repo directory and run "Setup Translation UI".
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/9417531/a0180268-be28-4d0d-aeda-d25fa861737c)
4. It will result in new UI:
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/9417531/0a513fbc-5996-4a41-b76b-7c4ce9d2315f)

-----

### Steps for users translating the addon
1. If `blenderbim-translations` already has has .po file for your language (e.g. `de_DE.po` for German), you can skip this step. Otherwise you need to create a folder in `blenderbim-translations` with your language code and copy `blenderbim.pot` to that folder and rename it according to your language (for German it would be `blenderbim.pot` -> `de_DE/de_DE.po`).

Currently supported languages at the moment are (the ones available in Blender UI): `'ca_AD', 'en_US', 'es', 'fr_FR', 'ja_JP', 'sk_SK', 'cs_CZ', 'de_DE', 'it_IT', 'ka', 'ko_KR', 'pt_BR', 'pt_PT', 'ru_RU', 'uk_UA', 'vi_VN', 'zh_HANS', 'zh_HANT', 'ab', 'ar_EG', 'bg_BG', 'el_GR', 'eo', 'eu_EU', 'fa_IR', 'fi_FI', 'ha', 'he_IL', 'hi_IN', 'hr_HR', 'hu_HU', 'id_ID', 'ky_KG', 'ne_NP', 'nl_NL', 'pl_PL', 'sr_RS', 'sr_RS@latin', 'sv_SE', 'sw', 'ta', 'th_TH', 'tr_TR'`.

2. Option 1 - edit .po files manually or with some dedicated software (as .po files are very popular format for storing translated strings, so some software might offer autotranslations or other useful features).
.po files by themselves have very simple format - they use original English messages as keys for the translation (`msgid`) and translated version of the string stored in `msgstr`.
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/9417531/5874b7b5-25a7-4dbe-9991-a5922d6f2bba)

2. Option 2 - edit strings from Blender UI. Downside of this method that not all strings can be reached that way - mainly just buttons and properties. You can use RMB on any button in the UI and it will have option to "Edit Translation".
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/9417531/dd163ab1-2735-47d7-971a-487ad8865215)

It will invoke a popup that will allow you edit the button label and it's tooltip translation. After typing in make sure "Save to PO file" is active (otherwise changes won't be saved to .po) and click OK.

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/9417531/477b498a-4654-4d4c-b165-33f64f8b7e58)

3. Update UI translations - it's available from BlenderBIM Translations tab in "Update Translations From .po".

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/9417531/6e02aa72-34d7-4044-a32b-f0b3ef170a98)

After update Blender will update `translations.py` and reload the translations in your UI and you'll be able to see the changes.
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/9417531/c5c5ace2-1f4a-43bb-bf1e-053fea9dcfab)

4. Commit / submit a PR to https://github.com/IfcOpenShell/blenderbim-translations with updated .po files.

-----

### Instructions for developers

The main source of the strings to translate for translators is "blenderbim.pot" file - it's simply .po template-file, it has the same format except it doesn't belong to any language in particular and all translated strings are empty. This file can be from BlenderBIM Translations UI using "Parse BlenderBIM strings to .po". Strings parsed in order as modules appear in BlenderBIM so it should be fairly consistent for diffs.

A bit details on `translations.py` - Blender reads translated strings from `translation_dict` stored in `translations.py` in addon directory. It is not present in https://github.com/IfcOpenShell/IfcOpenShell repo as this file might update too frequently and it's generated during BlenderBIM build using `bbim_translations.py` (the script is running completely without Blender or `bpy` module).

`translations.py` is just an intermediary format generated from .po files. It can be generated either from BlenderBIM Translation addon UI ("Update Translations From .po") or without blender `py `src/blenderbim/scripts/bbim_translations.py -i "C:/blenderbim-translations" -o "C:/Blender/4.0/scripts/addons/blenderbim/translations.py"`.
